bluesky/log.py

```diff
@@ -0,0 +1,157 @@
+        r"""
+        :arg bool color: Enables color support.
+        :arg str fmt: Log message format.
+          It will be applied to the attributes dict of log records. The
+          text between ``%(color)s`` and ``%(end_color)s`` will be colored
+          depending on the level if color support is on.
+        :arg dict colors: color mappings from logging level to terminal color
+          code
+        :arg str datefmt: Datetime format.
+          Used for formatting ``(asctime)`` placeholder in ``prefix_fmt``.
+        .. versionchanged:: 3.2
+           Added ``fmt`` and ``datefmt`` arguments.
+        """
+        super().__init__(datefmt=datefmt)
+        self._fmt = fmt
+
+        self._colors = {}
+        if color and _stderr_supports_color():
+            if curses is not None:
+                # The curses module has some str/bytes confusion in
+                # python3.  Until version 3.2.3, most methods return
+                # bytes, but only accept strings.  In addition, we want to
+                # output these strings with the logging module, which
+                # works with unicode strings.  The explicit calls to
+                # unicode() below are harmless in python2 but will do the
+                # right conversion in python 3.
+                fg_color = (curses.tigetstr("setaf") or
+                            curses.tigetstr("setf") or "")
+
+                for levelno, code in colors.items():
+                    self._colors[levelno] = str(curses.tparm(fg_color, code), "ascii")
+                self._normal = str(curses.tigetstr("sgr0"), "ascii")
+            else:
+                # If curses is not present (currently we'll only get here for
+                # colorama on windows), assume hard-coded ANSI color codes.
+                for levelno, code in colors.items():
+                    self._colors[levelno] = '\033[2;3%dm' % code
+                self._normal = '\033[0m'
+        else:
+            self._normal = ''
+
+    def format(self, record):
+        record.message = record.getMessage()
+        record.asctime = self.formatTime(record, self.datefmt)
+
+        try:
+            record.color = self._colors[record.levelno]
+            record.end_color = self._normal
+        except KeyError:
+            record.color = ''
+            record.end_color = ''
+
+        formatted = self._fmt % record.__dict__
+
+        if record.exc_info and not record.exc_text:
+            record.exc_text = self.formatException(record.exc_info)
+        if record.exc_text:
+            formatted = '{}\n{}'.format(formatted.rstrip(), record.exc_text)
+        return formatted.replace("\n", "\n    ")
+
+
+plain_log_format = "[%(levelname)1.1s %(asctime)s.%(msecs)03d %(module)s:%(lineno)d] %(message)s"
+color_log_format = ("%(color)s[%(levelname)1.1s %(asctime)s.%(msecs)03d "
+                    "%(module)s:%(lineno)d]%(end_color)s %(message)s")
+log_date_format = "%H:%M:%S"
```

Comment:
Can the date format be configurable here?

Reply:
I don't think there is a nice way to expose that option. This module code adds a handler to the 'bluesky' logger at import time. Some function that _mutates_ that handler (`set_log_date_format` or something) after the fact feels dirty to me. Better for the user to write their own formatter and handler, reusing some of the pieces here if they like -- in my opinion.

Reply:
OK, I am convinced this is actually a really good suggestion. Stand by for commit. :- D